### ISSUE TITLE
How to train under Multi-Source Closed-Set settings

### ISSUE THREAD

**Issue by e96031413** [silvia1993/HyMOS#1]
Hi @silvia1993, thanks for your contribution

I would like to know how to train HyMOS under Multi-Source Closed-Set settings to reproduce the Table 7 in the paper.

**Comment by silvia1993** [comment]
Hi @e96031413,

thank you for your interest!

We are working on the release of the code for that setting, we'll let you know as soon as it is available.

**Comment by e96031413** [comment]
@silvia1993 Thanks for your reply

**Comment by e96031413** [comment]
Hi @silvia1993,

I have a 1.3 million images custom dataset, which have 34 source domain and 1 target domain, the label set contains 6 label. (e.g. label 0, label 1, label 2, label 3, label 4, label 5)

But some of my source domain might contain only part of the label (e.g. "only label 0" or "label 0 and label 1 " or "label 0 and label 2"..........), and the target domain have the same situation.

To test if the HyMOS works for my dataset, I modified the datasets.py, so that the DataLoader can load these domains. However, the loading speed is slow for all these domains(1.3 million images). Hence, I load only 3 source domains that contain fewer images(only 26 images). and the code can work correctly.

Do you think I can still use HyMOS(Multi-Source Domain Adaptation method) for handle my problem?

or Do you know any other research direction is suitable for the problem setting?

The reason why I asked this question is that when I executed the code, "One dataset does not contain all the classes!" display on the screen for 28 times(Since 28 of my source domain doesn't contain all the classes)

**Comment by FrancescoCappio** [comment]
Hello!
The message "One dataset does not contain all the classes!" was added for debug purposes and, as you can probably notice, it is not an error and it should not stop the training. 
HyMOS was not designed for your particular setting as in all our benchmarks we considered situations in which all source domains share the same label set. We use this kind of prior knowledge when building training mini batches in order to insert in each mini batch one sample for each class from each source domain.  However it should still be possible to use HyMOS in your setting, maybe with some modifications. 
What I suggest is to look at the BalancedMultiSourceRandomSampler at the end of datasets.py. This is the module that creates sequences of samples used later to create batches. You should check that with your large datasets with source domains that do not share the label set the generated sequences are still correct and if they are not you should modify this module. 

Please take care that your final batches are still class-balanced and that, when possible, samples of the same class are collected from different domains. This will allow to maximize source-source alignment!

**Comment by e96031413** [comment]
@FrancescoCappio Thanks for your reply

**Comment by Anyway2022** [comment]
> Hi @e96031413,
> 
> thank you for your interest!
> 
> We are working on the release of the code for that setting, we'll let you know as soon as it is available.

Do you still have the plan for releasing this code?
